Exercise 4/behaviouralPattern/VisitorPattern.java

```diff
@@ -0,0 +1,85 @@
+package behaviouralPattern;
+
+public class VisitorPattern {
+	public static void main(String[] args) {
+		Child child=new Child();
+		Dog dog=new Dog();
+		child.setDog(dog);
+		child.playWithDog("stick");
+	}
+}
+
+class Child{
+	Dog dog;
+
+	public Dog getDog() {
+		return dog;
+	}
+
+	public void setDog(Dog dog) {
+		this.dog = dog;
+	}
+	
+	public void playWithDog(String item) {
+		try {
+			dog.play(item);
+		} catch (DogException e) {
+			e.printStackTrace();
+			e.visit();
+		}
+	}
+}
+
+class Dog{
+	public void play(String item) throws DogException{
+		if(item.equals("stick"))
+			throw new DogBarkException("dog is going to bark...");
+		else if(item.equals("stone"))
+			throw new DogBiteException("dog is going to bite...");
+		else
+			System.out.println("dog plays silently...");
+	}
+}
+
+abstract class DogException extends Exception{
+	public DogException(String msg) {
+		super(msg);
+	}
+	public abstract void visit();
```

Comment:
What is the use of abstract method here ? It seems to be visit() method has same logic everywhere. In such cases - Try not to repeat the code and have it in a single method

Reply:
I have used abstract method so that it can be implemented in derived classes. Since the code is repeating, I will try to remove the repeating code and place it in base class itself.

Reply:
I have moved the code of creating handler to base class
`handler=new Handler911();`

But handle works only if it is thrown from derived class since it uses the current instance using `this` keyword
```
public void visit() {
    handler.handle(this);
}
```